File: 02-two_pointers/00011-Container_with_Most_Water/00011-container_with_most_water.go
// 11: Container with Most Water
// https://leetcode.com/problems/container-with-most-water/

package main

import (
	"fmt"
	"math"
)

// SOLUTION
func maxArea(height []int) int {
	l := 0
	r := len(height) - 1
	result := 0

	for l<r {
		result = int(math.Max(float64(result), float64((r-l))*math.Min(float64(height[l]), float64(height[r]))))
		if height[l]<height[r] {
			l++
		} else {
			r--
		}
	}

	return result
}


func main() {
	// INPUT
	height := []int{1,8,6,2,5,4,8,3,7}

	// OUTPUT
	result := maxArea(height)
	fmt.Println(result)
}

File: 02-two_pointers/00011-Container_with_Most_Water/00011-container_with_most_water.java
// 11: Container with Most Water
// https://leetcode.com/problems/container-with-most-water/


class Solution {
    // SOLUTION
    public int maxArea(int[] height) {
        int l = 0;
        int r = height.length - 1;
        int result = 0;

        while (l<r) {
            result = Math.max(result, (r-l)*Math.min(height[l], height[r]));
            if (height[l]<height[r]) l++;
            else r--;
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] height = {1,8,6,2,5,4,8,3,7};

        // OUTPUT
        var result = o.maxArea(height);
        System.out.println(result);
    }
}